snapdtool/cmdutil.go

```diff
@@ -112,9 +112,27 @@ func CommandFromSystemSnap(name string, cmdArgs ...string) (*exec.Cmd, error) {
 	cmdPath := filepath.Join(root, name)
 
 	if from == "snapd" {
-		// We do not need to specify the linker or library
-		// path from files coming from the snapd snap.
-		return exec.Command(cmdPath, cmdArgs...), nil
+		// the elf interpreter invoked by the binary will work if snaps are mounted at /snap
+		// or /snap/snapd/current resolves to <mount dir>/snapd/current so that the interpreter
+		// locations are correct, otherwise we need to set up a command to invoke it directly
+		snapdCurrentDir := filepath.Join(dirs.GlobalRootDir, "snap/snapd/current")
+		if match, err := osutil.ComparePathsByDeviceInode(root, snapdCurrentDir); err == nil && match {
+			return exec.Command(cmdPath, cmdArgs...), nil
+		}
+
+		interp, err := elfInterp(cmdPath)
```

Comment:
But this would not work for static binaries, right? Maybe that will not happen for current binaries that we ship in snapd, just wondering.

Reply:
I don't think we call any other binary tham xdelta3/mksquashfs/unsquashfs using this function. In any case calling it through ld.so should be transparent as it will figure out that there is nothing to do.